### PR TITLE
#575, #1175 Show 'progress' indicator during verify/upload

### LINKED ISSUE #575: Show 'progress' indicator during verify/upload
**Is your feature request related to a problem? Please describe.**

There is no sign of activity from the IDE for long periods of time while compiling or uploading large sketches with the default setting of verbose output disabled.

**Describe the solution you'd like**

Add a visual progress indicator.

The classic Arduino IDE has this:

**Desktop (please complete the following information):**
- OS: Windows 10, Ubuntu 20.04
- Version: 2.0.0-beta.12-nightly.20211020
  Date: 2021-10-20T03:02:52.486Z
  CLI Version: 0.19.1 alpha [718bbbf2]

**Additional context**

Originally requested at:

- https://forum.arduino.cc/t/no-compilation-or-uploading-progress-bar-for-the-arduino-ide-2-beta/876983
- https://forum.arduino.cc/t/compile-progress-status-bar/908536
- https://forum.arduino.cc/t/ide-2-0-rc5-compile-progress-bar/977994

Related:

- https://github.com/arduino/arduino-ide/issues/279
- https://github.com/arduino/arduino-ide/pull/305

### LINKED ISSUE #1175: Output panel should open immediately at press of "verify" or "upload"
### Describe the request

Modify the behaviour of the output panel upon "verify" or "upload" actions being triggered

As requested by @tigoe in https://github.com/arduino/arduino-ide/issues/279 we should better communicate to the user the following as bare minimum:
* something is happening
* what is happening
  * verify start
  * verify success
  * upload start
  * upload success
* interleave these messages to the tools' standard output we'd see when "verbose" is on

### Describe the current behavior

The behaviour of the output panel differs from IDE 1.x as it won't open immediately at begin of "verify" or "upload" actions.
When compile/upload verbose is disabled, especially since the buttons are disabled at click, this can create the impression that nothing is happening until successful compilation (or error) output needs to be displayed triggering the opening of the output panel.

### Arduino IDE version

2.0 (all)

### Operating system

N/A

### Operating system version

n/a

### Additional context

_No response_

### Issue checklist

- [X] I searched for previous requests in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the feature was still missing when using the latest [nightly build](https://github.com/arduino/arduino-ide#nightly-builds)
- [X] My request contains all necessary details

### PR DESCRIPTION
**Pull request by kittaakos**
### Motivation

Show progress indication when running the `verify`, `upload`, `upload using programmer` or `burn bootloader` commands from IDE2.

The changes in action:
1. Verfiy:

https://user-images.githubusercontent.com/1405703/181473877-daa198b2-d102-446f-bf3d-1aa8d065e2e2.mp4

2. Upload:

https://user-images.githubusercontent.com/1405703/181473936-ba90256c-f925-4dde-9406-c9de7054af50.mp4

### Change description

 - [dev]: The core service (backend) does not run `compile` before `verify`; it's the caller's responsibility to run it.

### Other information

~Noteworthy UI/UX changes:~

~An `Upload` consists of two tasks: `Verfiy` and `Upload`. With the proposed changes, users will see on the toolbar (Verify toolbar icon is yellow with the light Arduino theme) that if `Verify` is in progress, then `Upload` is in progress. See in action.~

~https://user-images.githubusercontent.com/1405703/181474498-14e02b6b-1f4e-4b07-b93c-7e819f874e8e.mp4~

~This is a UI change compared to the `HEAD` of the `main` branch, but I believe this is the correct solution What do you think @per1234 and @ubidefeo?~

For reviewers:
 - The _Output_ view should be revealed but not focused on `Verify`/`Upload`,
 - Please double check #1132,
 - The progress indication must be the same for platform/library install/uninstall and index updates.

Closes #575
Closes #1175

### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: arduino-ide-extension/src/browser/widgets/component-list/component-list-item.tsx [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import * as React from '@theia/core/shared/react';
-import { Installable } from '../../../common/protocol/installable';
+import { Installable } from '../../../common/installable';
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/widgets/component-list/component-list-item.tsx:2]: TODO

File: arduino-ide-extension/src/node/grpc-progressible.ts [hunk 2]
```diff
@@ -78,20 +84,67 @@ namespace IndexProgressResponse {
...
 }
+namespace UnitOfWork {
+  export const Unknown: UnitOfWork = {};
+}
 
 /**
  * It's solely a dev thing. Flip it to `true` if you want to debug the progress from the CLI responses.
  */
-const DEBUG = false;
+const DEBUG = true;
```
Comment (kittaakos) [inline arduino-ide-extension/src/node/grpc-progressible.ts:147]: TODO: revert

**Comment by ubidefeo** [comment]
The progress indicator works as expected.
Opening the output panel right at the beginning of verify is helping a lot in moving the attention to the bottom of the window, where the progress can be noticed easy.
I have seen at the end of the ESP32 platform installation that a longer message appears in the output panel, I guess this is expected?

```
Installing esp32:riscv32-esp-elf-gcc@gcc8_4_0-esp-2021r2-patch3
esp32:riscv32-esp-elf-gcc@gcc8_4_0-esp-2021r2-patch3 installed
Installing esp32:xtensa-esp32-elf-gcc@gcc8_4_0-esp-2021r2-patch3
esp32:xtensa-esp32-elf-gcc@gcc8_4_0-esp-2021r2-patch3 installed
Installing esp32:xtensa-esp32s2-elf-gcc@gcc8_4_0-esp-2021r2-patch3
esp32:xtensa-esp32s2-elf-gcc@gcc8_4_0-esp-2021r2-patch3 installed
Installing esp32:xtensa-esp32s3-elf-gcc@gcc8_4_0-esp-2021r2-patch3
esp32:xtensa-esp32s3-elf-gcc@gcc8_4_0-esp-2021r2-patch3 installed
Installing esp32:esptool_py@3.3.0
esp32:esptool_py@3.3.0 installed
Installing esp32:mkspiffs@0.2.3
esp32:mkspiffs@0.2.3 installed
Installing esp32:mklittlefs@3.0.0-gnu12-dc7f933
esp32:mklittlefs@3.0.0-gnu12-dc7f933 installed
Installing platform esp32:esp32@2.0.3
Configuring platform.
Platform esp32:esp32@2.0.3 installed
```

cc @per1234

**Review by AlbyIanna — APPROVED** [review]
The code looks good, but I have only one remark.

I noticed that it's possible to trigger an Upload when there's a Verify in progress, and vice versa it's possible to trigger a Verify when there's an Upload in progress. Maybe this is expected (in the end it doesn't seem to cause any significant issue). I'm pointing this out because it feels weird to click on the Upload button and get feedback (the yellow active status) on the Verify button, while the Upload button is still clickable.

**Comment by kittaakos** [comment]
Thanks for taking the time and trying it out.

> I noticed that it's possible to trigger an Upload when there's a Verify in progress, and vice versa it's possible to trigger a Verify when there's an Upload in progress. Maybe this is expected (in the end it doesn't seem to cause any significant issue)

It's the same in the `main` too. So this did not change anything.

Screen Shot 2022-07-29 at 17 29 28


> I'm pointing this out because it feels weird to click on the Upload button and get feedback (the yellow active status) on the Verify button

Good that you brought this up; it's mentioned in the description as expected behavioral changes. I am looking forward to hearing how it should exactly work. I can restore what it is on the `main`, but you also noticed the behavior does not seem correct, right?

> while the Upload button is still clickable.

This is undesired. Thanks for noting it. I will change the logic so that when the user clicks on `Upload`, the following will happen:
 - upload turns into yellow,
 - verify starts (verify button yellow, upload is still yellow),
 - verify completes (only upload is yellow),
 - upload completes (zero yellow buttons)

**Comment by AlbyIanna** [comment]
> I am looking forward to hearing how it should exactly work. 

I don't really know. The Java IDE didn't have this issue by turning yellow only one button: if you clicked on the Upload button, the Upload button is the only one that turns yellow (even during the compile process), and clicking the Verify button during this time wouldn't do anything, making it disabled for all intents. But anyway, as you said, it's not in this PR's scope to solve this issue 👍  


> upload turns into yellow,
> verify starts (verify button yellow, upload is still yellow),
> verify completes (only upload is yellow),
> upload completes (zero yellow buttons)

I like this, thanks!

Let's add @91volt to the thread and see what he thinks about this solution.

**Comment by kittaakos** [comment]
> > I am looking forward to hearing how it should exactly work.
> 
> I don't really know. The Java IDE didn't have this issue because it turns yellow only one button: if you click on the Upload button, the Upload button is the only one that turns yellow (even during the compile process), and clicking the Verify button during this time wouldn't do anything, making it disabled for all intents. But anyway, as you said, it's not in this PR's scope to solve this issue 👍
> 
> > upload turns into yellow,
> > verify starts (verify button yellow, upload is still yellow),
> > verify completes (only upload is yellow),
> > upload completes (zero yellow buttons)
> 
> I like this, thanks!

I reverted all the UX changes. The `Verify` and `Upload` UX should be the same as the `main` state.

**Review by per1234 — APPROVED** [review]
This is a really nice advancement in the user experience. Thanks Akos!

**Comment by per1234** [comment]
>I noticed that it's possible to trigger an Upload when there's a Verify in progress, and vice versa it's possible to trigger a Verify when there's an Upload in progress.

This bug is now being tracked at https://github.com/arduino/arduino-ide/issues/1299

**Pull request merged**